### ISSUE TITLE
Tracking issue for iter_order_by

### ISSUE THREAD

**Issue by KodrAus** [rust-lang/rust#64295]
Landed in #62205

Before stabilization:

- [ ] Accepted RFC
- [ ] Stabilization PR

**Comment by KodrAus** [comment]
A [comment was made on Twitter](https://twitter.com/BarryRevzin/status/1350861720185663488) about whether or not this should be able to use `size_hint` to determine whether or not the iterators have any chance of containing equal elements.

`size_hint` can't be relied upon for safety, but it seems like it could be reasonable to use it for short-circuiting `eq_by`.

**Comment by yoshuawuyts** [comment]
An example usage of this is in [rust-analyzer](https://github.com/rust-analyzer/rust-analyzer/blob/49dbb74a0b4af13e8c83c667fcddb6ff1a931c8d/crates/stdx/src/lib.rs#L149) where it's part of its internal `stdx` crate which contains stdlib extensions.

**Comment by yoshuawuyts** [comment]
> `size_hint` can't be relied upon for safety, but it seems like it could be reasonable to use it for short-circuiting `eq_by`.

Perhaps this could be specialized in the case `I::Iterator: ExactSizeIterator` [(ref)](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html)? That way the base case can remains correct regardless of the value of `size_hint`, but if we know the length we can short-circuit.

**Comment by chrysn** [comment]
TrustedLen is an unsafe trait, which I understand is there to catch cases when wrong lengths would result in soundness holes. The size hint is a bit more than a hint -- an iterator that does not adhere to the bounds is considered buggy by [size_hint's documentation](https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.size_hint). I'd place returning false to `eq_by` on an erroneous iterator in a similar category as a run-time integer overflow that is not caught because the wrong index is not a soundness violation (just a follow-up program error from a previous program error).

Is that the only contested point about this interface? If not, what is needed to move this toward stabilization? (If it's user reports: it simplifies code in coap-handler-implementations quite a bit).

**Comment by Hiten-Tandon** [comment]
Function should be renamed to related_by, because, in the function we're not checking for equality, rather we're checking for relation between the values of the two iterators.

**Comment by orlp** [comment]
So is there any progress on this? What's blocking stabilization?

**Comment by leb-kuchen** [comment]
I think the question should be TrustedLen or ExaxtSizeIterator. ExaxtSizeIterator is also user implemented, but is supposed to be exact and not only a hint.

**Comment by anatawa12** [comment]
I agree with orlp, I think there's no blocker for this unstable feature.

There is an issue about naming eq_by in #100795 but no update and I feel it's more sensible to add alias in another feature gate or keep as is since `eq_by` is symmetrical to `eq` and other _by functions.
(This is my opinion so needs libs-api team decision)

Therefore I feel this feature is ready for FCP.

@rust-lang/libs-api Could you start FCP for long feature?

It looks this tracking issue lacks implementation history and Experience Report for stabilization report so I'm writing here.

-----

## Implementation History

- Initial implementation #62205 and #64296
  - Please note that PR description says {ne,lt,le,gt,ge}_by are added, but they are removed before merge. Remaining ones are `cmp_by`, `partial_cmp_by`, and `eq_by` which are exactly same as current ones
- Refactoring and performance improvements with ControlFlow #100845

## Experience Report

- There is usage in rust-analyzer https://github.com/rust-lang/rust-analyzer/blob/f3998f7f8a197596c5edf72e937996e6674b423b/crates/stdx/src/lib.rs#L275-L301

**Comment by leb-kuchen** [comment]
For `Ord` there are the methods on [Ordering](https://doc.rust-lang.org/stable/std/cmp/enum.Ordering.html#implementations). For `PartialOrd`, Option::is_some_and can be used. 
I don't see any specializations for {ne,lt,le,gt,ge}. Thus I don't think these methods justify the bloat.

**Comment by anatawa12** [comment]
{lt,le,gt,ge} functions are not part of this feature flag and already stable since 1.5.0 so removing those functions are generally impossible and it's out of scope of this feature I think.

**Comment by yotamofek** [comment]
`eq_by` is really cool (see #147101). Just came here to say that `eq_by` might not be the best name, but I see that it's been said, so +1 on that. (See here: https://github.com/rust-lang/rust/pull/147101/files#diff-593003090e0fb5c21f31413ce5feb506e235ec33c4775da88b853980429b9ff1R2806 for example of where `eq` is just the wrong name for this).

I'd also like to see if anyone else sees benefit in adding a `prefix_eq` family of functions, i.e. the same as `eq` but where one iterator being shorter or longer than the other doesn't imply inequality. (So basically, the same as `iter::zip(a, b).all(|(a, b)| a == b)`). I think it can be useful for making code more readable.

**Comment by yotamofek** [comment]
> So is there any progress on this? What's blocking stabilization?
> 
> Regarding the size hint being used, that's currently not being done for the **already stable** `Iterator::eq`. I think there are really only two paths forward:
> 
>     1. Don't check the size hint.
> 
>     2. Check the size hint in all `Iterator` `PartialEq`-related methods. This means deciding it's not considered a breaking change to start doing this for `Iterator::eq`.
> 
> 
> I really don't think there is ever a viable future for having _some_ of the methods check the size hint and others do not, so I really believe the two above choices are the only ones. It's just too much of a footgun.
> 
> However regardless of which of these two choices you make, `Iterator::eq_by`'s stabilization is not blocked by it. In scenario 1 the current implementation is fine. In scenario 2 you've already decided it's not a breaking change so it can be stabilized.

I didn't notice there was a question about specialization on `TrustedLen` here, but I ended up doing that anyways for both `eq` and `eq_by` here: #137122

**Comment by tisonkun** [comment]
This can be related to https://github.com/rust-lang/libs-team/issues/527 where the `by` param can be a `Compare` trait, while we have a default `impl<L, R> Compare<L, R> for FnMut(L, R) -> Ordering`.
